{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/grosa-isonlin3/stunning-octo-chainsaw/blob/main/bioinformaticstoolbox.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "This is our Bioinformatics ToolKit to use if you start personal project: Functions [1-9]"
      ],
      "metadata": {
        "id": "sx7LknewGYqJ"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wOfOV6w1cWt0",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "57e64138-77c4-4534-edd6-71889373fb5a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Type your DNA Strand here:CAGCAGCA\n",
            "[1] Yay, this is your DNA Strand:CAGCAGCA\n",
            "[2] Here is a randomly generated DNA: AGGTAAAGTTTCAGCATTAG\n",
            "[3] DNA Strand:AGGTAAAGTTTCAGCATTAG\n",
            "{'A': 7, 'C': 2, 'G': 5, 'T': 6}\n",
            "[4] Nucleotide Frequency Length: {20}\n",
            "[5] DNA/RNA Transcription: AGGUAAAGUUUCAGCAUUAG\n",
            "[6] This is our complementary strand: CTAATGCTGAAACTTTACCT\n",
            "[7] Making a Actual DNA strand with Reverse Comple + DNA: \n",
            "5' AGGTAAAGTTTCAGCATTAG 3'\n",
            "   ||||||||||||||||||||\n",
            "3' CTAATGCTGAAACTTTACCT 5' \n",
            "[8] RNA Strand:AGGUAAAGUUUCAGCAUUAG\n",
            "{'A': 7, 'C': 2, 'G': 5, 'U': 6}\n",
            "[9]This is our DNA with Colors \u001b [92nA\u001b [93nG\u001b [93nG\u001b [91nT\u001b [92nA\u001b [92nA\u001b [92nA\u001b [93nG\u001b [91nT\u001b [91nT\u001b [91nT\u001b [94nC\u001b [92nA\u001b [93nG\u001b [94nC\u001b [92nA\u001b [91nT\u001b [91nT\u001b [92nA\u001b [93nG\n",
            "[10] This is our GC content in DNA Strand: 35.0 %\n",
            "[11] This is our GC content at k=5 [40.0, 20.0, 60.0, 20.0]\n"
          ]
        }
      ],
      "source": [
        "\n",
        "DNANucleotides = ['A', 'C','G','T']\n",
        "ReverseComp = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
        "dna_seq = input(\"Type your DNA Strand here:\") # You input your DNA strand\n",
        "DNAString = \"TTTCACAGCGATCGCGCGTTTTACACTAC\" # Sample DNA String\n",
        "# Function 1: Validating a sequence if its true ------------------------------------\n",
        "def ValidateSeq(dna_seq):\n",
        "  tmpseq = dna_seq.upper()\n",
        "  for nuc in tmpseq:\n",
        "      if nuc not in DNANucleotides:\n",
        "        return print(\"Sorry that's not a DNA Strand :(\")\n",
        "        break\n",
        "  return tmpseq\n",
        "print(\"[1] Yay, this is your DNA Strand:\" + ValidateSeq(dna_seq)) #\n",
        "\n",
        "# Function 2 Random DNA Generator --------------------------------------------------\n",
        "import random\n",
        "randDNAstr = ''.join([random.choice(DNANucleotides) for nuc in range(20)])\n",
        "print(\"[2] Here is a randomly generated DNA: \" + ValidateSeq(randDNAstr))\n",
        "DNAstring = ValidateSeq(randDNAstr)\n",
        "\n",
        "# Function 3 Counting Nucleotides in DNA ---------------------------------------------------\n",
        "def countNucfrequencyDNA(seq):\n",
        "  temporaryFrequencyDictionary = {\"A\": 0, \"C\": 0, \"G\": 0, \"T\": 0}\n",
        "  for nuc in DNAstring:\n",
        "    temporaryFrequencyDictionary[nuc] += 1\n",
        "  return temporaryFrequencyDictionary\n",
        "print(\"[3] DNA Strand:\" + ValidateSeq(randDNAstr))\n",
        "print(countNucfrequencyDNA(DNAstring))\n",
        "print(\"[4] Nucleotide Frequency Length:\", {len(DNAstring)})\n",
        "\n",
        "# Function 4 Transcription DNA -> RNA ------------------------------------------------------\n",
        "def transcription(seq):\n",
        "  # Replaces Thymine with Uracil in DNA\n",
        "  return seq.replace(\"T\", \"U\")\n",
        "print(\"[5] DNA/RNA Transcription: \" + transcription(DNAstring)) # Turned DNA -> RNA\n",
        "RNAstring = transcription(DNAstring)\n",
        "\n",
        "# Function 5 Reverse Complementary DNA Strand --------------------------------\n",
        "def reversecomplement(seq):\n",
        "  mapping = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
        "  return \"\".join([mapping[base] for base in seq[::-1]])\n",
        "print(\"[6] This is our complementary strand:\", reversecomplement(DNAstring))\n",
        "\n",
        "\n",
        "# Putting Reverse Complementary and Original STrand together -------\n",
        "print(\"[7] Making a Actual DNA strand with Reverse Comple + DNA: \")\n",
        "print(\"5' \" + (DNAstring) + \" 3'\")\n",
        "print( '   ' + ''.join(['|' for c in range(len(DNAstring))]))\n",
        "print(\"3' \" + reversecomplement(DNAstring) + \" 5' \")\n",
        "\n",
        "# Function 6 Counting Nucleotides in RNA --------------------------------------\n",
        "def countNucfrequencyRNA(seq):\n",
        "  temporaryFrequencyDictionary = {\"A\": 0, \"C\": 0, \"G\": 0, \"U\": 0}\n",
        "  for nuc in RNAstring:\n",
        "    temporaryFrequencyDictionary[nuc] += 1\n",
        "  return temporaryFrequencyDictionary\n",
        "print(\"[8] RNA Strand:\" + RNAstring)\n",
        "print(countNucfrequencyRNA(RNAstring))\n",
        "\n",
        "\n",
        "# Function 7 Colorizing DNA Ouput for Presentation (Python Preference) ---------------------------\n",
        "def colored_montified(seq):\n",
        "  bcolors = {\n",
        "      'A': '\\033 [92n',\n",
        "      'C': '\\033 [94n',\n",
        "      'G': '\\033 [93n',\n",
        "      'T': '\\033 [91n',\n",
        "      'U': '\\033 [91n',\n",
        "      'reset': '\\033 [0;0n'\n",
        "  }\n",
        "  tmpStr = ' '\n",
        "  for nuc in seq:\n",
        "    if nuc  in bcolors:\n",
        "      tmpStr += bcolors[nuc] + nuc\n",
        "    else:\n",
        "      tmpStr += bcolors['reset'] + nuc\n",
        "  return tmpStr\n",
        "print(\"[9]This is our DNA with Colors\" + colored_montified(DNAstring))\n",
        "# This will only show up in the terminal code, You write \"python filename.py\"\n",
        "\n",
        "\n",
        "# Function 8 GC Content Calculations ----------------------------------------\n",
        "def gc_content (seq):\n",
        "    return round ((seq.count(\"C\") + seq.count(\"G\")) / len(seq) * 100, 6)\n",
        "print(\"[10] This is our GC content in DNA Strand:\",gc_content(DNAstring), \"%\")\n",
        "\n",
        "# Function 9 GC Content Calculation in Specific Window ------------------------\n",
        "def gc_content_subseq(seq, k):\n",
        "    res = [] # Our result stored here\n",
        "    for i in range(0, len(seq) - k + 1, k):\n",
        "        subseq = seq[i:i + k]\n",
        "        res.append(gc_content(subseq))\n",
        "    return res\n",
        "\n",
        "print(\"[11] This is our GC content at k=5\", gc_content_subseq(DNAstring, k=5))\n",
        "#We can give K a value we want, its how much character we want to calculate our DNAstring by."
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Bioinformatics Toolkit Functions [12 - continued...]:"
      ],
      "metadata": {
        "id": "TWkF7-N6GHzG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from collections import Counter\n",
        "#---------------------------- CODONS & Nucleotide DICTIONARIES: ---------------\n",
        "# Chart with all DNA codons used for Translation\n",
        "DNA_Codons = {\n",
        "    # 'M' - START, '_' - STOP\n",
        "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
        "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
        "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
        "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
        "    \"ATA\": \"I\", \"ATT\": \"I\", \"ATC\": \"I\",\n",
        "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
        "    \"ATG\": \"M\",\n",
        "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
        "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n",
        "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
        "    \"TGG\": \"W\",\n",
        "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
        "    \"TAA\": \"_\", \"TAG\": \"_\", \"TGA\": \"_\"\n",
        "}\n",
        "\n",
        "NUCLEOTIDE_BASE = {\n",
        "    \"DNA\": [\"A\", \"T\",\"C\", \"G\"],\n",
        "    \"RNA\": [\"A\", \"U\",\"C\", \"G\"],\n",
        "}\n",
        "\n",
        "RNA_Codons = {\n",
        "    # 'M' - START, '_' - STOP\n",
        "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "    \"UGU\": \"C\", \"UGC\": \"C\",\n",
        "    \"GAU\": \"D\", \"GAC\": \"D\",\n",
        "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "    \"UUU\": \"F\", \"UUC\": \"F\",\n",
        "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "    \"CAU\": \"H\", \"CAC\": \"H\",\n",
        "    \"AUA\": \"I\", \"AUU\": \"I\", \"AUC\": \"I\",\n",
        "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"UUA\": \"L\", \"UUG\": \"L\", \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
        "    \"AUG\": \"M\",\n",
        "    \"AAU\": \"N\", \"AAC\": \"N\",\n",
        "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\", \"AGU\": \"S\", \"AGC\": \"S\",\n",
        "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
        "    \"UGG\": \"W\",\n",
        "    \"UAU\": \"Y\", \"UAC\": \"Y\",\n",
        "    \"UAA\": \"_\", \"UAG\": \"_\", \"UGA\": \"_\"\n",
        "}\n",
        "\n",
        "\n",
        "\n",
        "# [12] Translation into Amino Acids Function -------------------------------\n",
        "def translate_seq(seq, init_pos=0):\n",
        "    return [DNA_Codons[seq[pos:pos + 3]] for pos in range(init_pos, len(seq) - 2, 3)]\n",
        "print(f'[12] Aminoacid Sequences from DNA: {translate_seq(DNAstring, 0)} \\n')\n",
        "\n",
        "# [13] Counts the Codons/Amino acid usage in specific DNA sequeunce --------------------\n",
        "def codon_usage(seq, aminoacid):\n",
        "    tmplist = []\n",
        "    for i in range(0, len(seq) - 2, 3):\n",
        "        if DNA_Codons[seq[i:i + 3]] == aminoacid:\n",
        "            tmplist.append(seq[i:i + 3])\n",
        "    freqDict = dict(Counter(tmplist))\n",
        "    totalWight = sum(freqDict.values())\n",
        "    for seq in freqDict:\n",
        "        freqDict[seq] = round(freqDict[seq] / totalWight, 2)\n",
        "    return freqDict\n",
        "print(f'[13] This counts the amino acids of (\"Specific Amino Acid\"): {codon_usage(DNAstring, \"V\")}\\n')\n",
        "# You can change the letter depending on which amino acid you want to change _________________^\n",
        "\n",
        "\n",
        "\n",
        "# [14] Generating Reading Frames for Proteins in FUture --------------------------------\n",
        "def gen_reading_frames(seq):\n",
        "    frames = []\n",
        "    frames.append(translate_seq(seq, 0)) # Start from reading from start of frame\n",
        "    frames.append(translate_seq(seq, 1)) # THis function works with 5 and\n",
        "    frames.append(translate_seq(seq, 2)) # Three reading frames\n",
        "    frames.append(translate_seq(reversecomplement(seq), 0))\n",
        "    frames.append(translate_seq(reversecomplement(seq), 1))\n",
        "    frames.append(translate_seq(reversecomplement(seq), 2))\n",
        "    return frames\n",
        "    #We have reversecomplement as well to complete the DNA\n",
        "\n",
        "print(\"[14] This is our Reading Frames:\")\n",
        "for frame in gen_reading_frames(DNAstring):\n",
        "    print(frame)\n",
        "\n",
        "\n",
        "# [15] Protein Stop and Start Codon Analysis -------------------------------------\n",
        "def protein_reading_analysis(aa_seq):\n",
        "  # M is our start codon, _ is out stop codon according to our dictionary\n",
        "  current_prot = []\n",
        "  proteins = [] # <- Proteins added which are those after every stop codon\n",
        "  for aa in aa_seq:\n",
        "    if aa == \"_\": # <- Our current codon being searched\n",
        "      # This Part searches for Start Codons\n",
        "      if current_prot:\n",
        "        for p in current_prot:\n",
        "          proteins.append(p)\n",
        "        current_prot = []\n",
        "    else:\n",
        "      # This Part searches for Stop Codons\n",
        "      if aa == \"M\":\n",
        "        current_prot.append(\"\")\n",
        "    for i in range(len(current_prot)):\n",
        "      current_prot[i] += aa\n",
        "  return proteins\n",
        "test_amino_acids = ['T', 'M', 'H', 'I', 'K', 'T', 'L', 'H', 'G', 'L'\n",
        "   'P', 'H', 'T', 'S', 'R', 'L', 'Y', 'T', '_']\n",
        "print(\"\\n[15] Here is our protein using Stop & Start Codons:\")\n",
        "print(protein_reading_analysis(test_amino_acids))\n",
        "\n",
        "\n",
        "# [16] --- POINT MUTATIONS DETECTION ----------------------------------\n",
        "# For this to work, we need two strings at same length, Example Below:\n",
        "DNAString1 = \"GAGCCTACTAACGGGAT\"\n",
        "DNAString2 = \"CATCGTAATGACGGCCT\"\n",
        "# This function tracks how many mutations are there in the String\n",
        "def hamming_distance_pointmutation (DNAString1, DNAString2):\n",
        "  if len(DNAString1) != len(DNAString2):\n",
        "    print(\"Strings are not equal size, you can't detect mutations\")\n",
        "  else:\n",
        "    hamming_distance = 0\n",
        "    for i in range(len(DNAString1)):\n",
        "      if DNAString1[i] != DNAString2[i]:\n",
        "        hamming_distance += 1\n",
        "    return hamming_distance\n",
        "print(f'[16]The amount of Point Mutations detected on two DNA strands are: {hamming_distance_pointmutation(DNAString1, DNAString2)}')\n",
        "\n",
        "\n",
        "# [17] All Proteins in Open Reading Frames\n",
        "def all_proteins_in_orfs (seq, startReadPos=0, endReadPos=0, ordered=False):\n",
        "  # This function needs functions 12,14,and 15 to run\n",
        "  if endReadPos > startReadPos:\n",
        "    rfs = gen_reading_frames(seq[startReadPos:endReadPos])\n",
        "  else:\n",
        "    rfs = gen_reading_frames(seq)\n",
        "\n",
        "  #Finds proteins\n",
        "  res = []\n",
        "  for rf in rfs:\n",
        "    proteins = protein_reading_analysis(rf)\n",
        "    for protein in proteins:\n",
        "        res.append(protein)\n",
        "  if ordered:\n",
        "    return sorted(res, key=len, reverse=True)\n",
        "  return res\n",
        "\n",
        "  # Executes our function:\n",
        "print(f'[17] All Proteins in ORFS Printed:')\n",
        "for prot in all_proteins_in_orfs(DNAString, 0, 0, True):\n",
        "  print(f\"{prot}\")\n",
        "from os import write\n",
        "\n",
        "# [18] READS A TEXTFILE ----------------------------------------------------\n",
        "def readTextFile(filePath):\n",
        "  with open(filepath, 'r') as f:\n",
        "    return \"\".join([l.strip() for l in f.readlines()])\n",
        "   # reads file, line by line and strips any new lines\n",
        "\n",
        "\n",
        "# [19] WRITES A TEXTFILE --------------------------------------------------\n",
        "def writeTextFile(filePath, seq, mode='w'):\n",
        "  with open(filepath, mode) as f:\n",
        "    f.write(seq + '\\n')\n",
        "    #Overwrites our file. Makes a file with output\n",
        "\n",
        "#writeTextFile(\"test.txt, test_dna.seq\")\n",
        "#for rf in test_dna.gen_reading_frame():\n",
        "  #writeTextFile(\"test.txt\", str(rf), 'a')\n",
        "\n",
        "# [20] READS A FASTAFILE -----------------------------------\n",
        "def read_FASTA(filePath):\n",
        "  with open(filePath, \"r\") as f:\n",
        "    FASTAfile = [l.strip() for l in f.readlines()]\n",
        "  FASTADic = {}\n",
        "  FASTAlabel = \"\"\n",
        "  for line in FASTAfile:\n",
        "    if \">\" in line:\n",
        "      FASTAlabel = line[1:]\n",
        "      FASTADic[FASTAlabel] = \"\"\n",
        "    else:\n",
        "      FASTADic[FASTAlabel] += line\n",
        "  return FASTADic\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "psMXj9lLGlgF",
        "outputId": "396061dd-02b0-4003-fbc5-6bbc01dbb8e5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[12] Aminoacid Sequences from DNA: ['R', '_', 'S', 'F', 'S', 'I'] \n",
            "\n",
            "[13] This counts the amino acids of (\"Specific Amino Acid\"): {}\n",
            "\n",
            "[14] This is our Reading Frames:\n",
            "['R', '_', 'S', 'F', 'S', 'I']\n",
            "['G', 'K', 'V', 'S', 'A', 'L']\n",
            "['V', 'K', 'F', 'Q', 'H', '_']\n",
            "['L', 'M', 'L', 'K', 'L', 'Y']\n",
            "['_', 'C', '_', 'N', 'F', 'T']\n",
            "['N', 'A', 'E', 'T', 'L', 'P']\n",
            "\n",
            "[15] Here is our protein using Stop & Start Codons:\n",
            "['MHIKTLHGLPHTSRLYT']\n",
            "[16]The amount of Point Mutations detected on two DNA strands are: 7\n",
            "[17] All Proteins in ORFS Printed:\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here are all the Functions in the Geonome Toolkit:"
      ],
      "metadata": {
        "id": "TsQBwdtXXrrX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class genomeToolkit:\n",
        "  # [G21] To start a virtual envrionment\n",
        "  def __init__(self):\n",
        "    print(\"Genome Toolkit Initated\")\n",
        "  # [G22]\n",
        "  def count_kmer(self, sequence, kmer):\n",
        "    # counts repeating kmers in a sequence. Include overlapping kmers\n",
        "    kmer_count = 0\n",
        "    for i in range(len(sequence)-len(kmer)+1):\n",
        "      if sequence[i:i+len(kmer)] == kmer:\n",
        "        kmer_count += 1\n",
        "    return kmer_count\n",
        "\n",
        "\n",
        "# Execution -------------------------------------\n",
        "print(f\"[G21] Creates Virutal Environment:\",genomeToolkit())\n",
        "gt = genomeToolkit()\n",
        "seq = \"AAAGAAAATTGA\"\n",
        "kmer = \"AA\"\n",
        "print(f\"[G22] The current kmer count is:\", gt.count_kmer(seq, kmer))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WPE1N3FEsCtV",
        "outputId": "a59ccf39-2d40-44fe-d944-e77cd02161dd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Genome Toolkit Initated\n",
            "[G21] Creates Virutal Environment: <__main__.genomeToolkit object at 0x7d9760220dd0>\n",
            "Genome Toolkit Initated\n",
            "[G22] The current kmer count is: 5\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here are all the Functions used to Solve Rosalind Problems: (Note: They are commented, so remove the comments to run code)"
      ],
      "metadata": {
        "id": "qmgbtIkhBXQY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# Problem 1: Data Processing a FASTA File to get highest GC Content ---------------------\n",
        "# ------------------------------------------------------------------------------\n",
        "  # Note: File must be downloaded from Rosalind in order to work\n",
        "          # Leetcode Rules for this COde:\n",
        "            #1. Read from FATSA File format\n",
        "            #2. Clean/Prepare the data\n",
        "            #3. format the dataD\n",
        "            #4. run operation\n",
        "\n",
        "## ------------  STEP 1 READ FASTA FILE FORMAT\n",
        "# Reads a file and gives a list of lines/dictionaries\n",
        "#def readFile (filepath):\n",
        "  #with open (filepath, 'r') as f:\n",
        "    #return [l.strip()for l in f.readlines()] # this is a List comprehension\n",
        "    #that returns a list in FASTAfile.\n",
        "# Gets GC Content\n",
        "#def gc_content (seq):\n",
        "    #return round ((seq.count(\"C\") + seq.count(\"G\")) / len(seq) * 100, 6)\n",
        "\n",
        "# variables to store data:\n",
        "#variable to store file data\n",
        "#FASTAFile = readFile (\"test_data/gc_content.txt\") # <- You can switch this depending on File name\n",
        "# Dictionary for labels plus data\n",
        "#FASTADict = {}\n",
        "# String to hodl current label\n",
        "#FASTAlabel = \"\"\n",
        "## --------------------------- STEP 2 CLEAN DATA\n",
        "# Converting FASTAfile into a dictionary\n",
        "#for line in FASTAFile:\n",
        "  #if \">\" in line:\n",
        "    #FASTAlabel = line\n",
        "    #FASTADict[FASTAlabel] = \"\"\n",
        "  #else:\n",
        "    #FASTADict[FASTAlabel] += line\n",
        "# ------------ STEP 3 FORMAT THE DATA & Run GC Content\n",
        "# Uses dictionary comprehension to show a dictionary result of Each label and GC\n",
        "#RESULTDic = {key: gc_content(value) for (key, value) in FASTADict.items()}\n",
        "# Gets max value of highest GC in our file:\n",
        "#MAXGCvalue = max(RESULTDic, key=RESULTDic.get)\n",
        "#  ------------ STEP 4 COLLECT THE RESULTS\n",
        "#print(f'{MAXGCvalue[1:]}\\n{RESULTDic[MAXGCvalue]}') # Prints value for Rosalind\n",
        "# Shift + Alt + A to uncomment everything"
      ],
      "metadata": {
        "id": "77_dEBgxBWyB"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}